libtenzir/builtins/operators/throttle.cpp

```diff
@@ -0,0 +1,162 @@
+        = duration_cast<float_seconds>(now - last_timestamp).count()
+          * bandwidth_per_second_;
+      budget = std::min(bytes_per_window, budget + additional_budget);
+      auto split_position = static_cast<size_t>(budget);
+      auto head = chunk::make_empty();
+      auto tail = chunk::make_empty();
+      std::tie(head, tail) = split_chunk(bytes, split_position);
+      budget -= static_cast<double>(head->size());
+      co_yield std::move(head);
+      // If we didn't have enough budget to send everything in one go,
+      // send the remainder in intervals according to the configured
+      // window.
+      while (tail->size() > 0) {
+        budget = 0;
+        ctrl.set_waiting(true);
+        ctrl.self()
+          .request(alarm_clock, caf::infinite,
+                   duration_cast<caf::timespan>(window_))
+          .await(
+            [&]() {
+              ctrl.set_waiting(false);
+            },
+            [&ctrl](const caf::error& err) {
+              diagnostic::error("throttle operator failed to delay")
+                .note("encountered error: {}", err)
+                .emit(ctrl.diagnostics());
+            });
+        std::tie(head, tail)
+          = split_chunk(tail, static_cast<size_t>(bytes_per_window));
+        co_yield {}; // Await the alarm clock.
+        co_yield std::move(head);
+      }
+      last_timestamp = std::chrono::system_clock::now();
+    }
+    co_return;
+  }
+
+  auto name() const -> std::string override {
+    return "throttle";
+  }
+
+  auto optimize(expression const& filter,
+                event_order order) const -> optimize_result override {
+    (void)filter, (void)order;
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, throttle_operator& x) -> bool {
+    return f.apply(x.bandwidth_per_second_);
+  }
+
+private:
+  double bandwidth_per_second_ = 0.0; // bytes
+  float_seconds window_ = {};
+};
+
+class throttle_plugin final : virtual public operator_plugin<throttle_operator>,
+                              operator_factory_plugin {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto parse_operator(parser_interface& p) const -> operator_ptr override {
+    auto const* docs = "https://docs.tenzir.com/operators/throttle";
+    auto parser = argument_parser{"throttle", docs};
+    auto max_bandwidth = std::optional<uint64_t>{};
+    // TODO: Add option to set window size.
+    auto window = float_seconds{1};
+    parser.add(max_bandwidth, "<max_bandwidth>");
+    parser.parse(p);
+    if (not max_bandwidth) {
+      diagnostic::error("`max_bandwidth` must be a numeric value")
+        .note("the unit of measurement is bytes/second")
+        .throw_();
+    }
+    return std::make_unique<throttle_operator>(*max_bandwidth, window);
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    auto max_bandwidth = std::optional<uint64_t>{};
```

Comment:
I think you may want to make this a `located<uint64_t>`. Not optional so that it is actually required, and located so that you can highlight the location with `.primary(bandwidth.source)` in diagnostics.

Reply:
As far as I can tell `argument_parser2` only supports optional for the `.add(name, ...)` API?

Reply:
There's `.add("name", var)` and `.add(var, "name")`, one for positional and one for named arguments. Right now, named arguments must be optional.

Reply:
How do you want this operator to be used? If you want to go for `throttle events=42` and `throttle bytes=42` (to enable both for when the input type is events), then you want `std::optional<located<uint64_t>>` . If you want a positional argument instead, then you need `.add(..., "<whatever>")`. Required positionals could also come soon, but right now you

Reply:
I think I'd prefer `throttle 42, over=10s`.

Reply:
Matthias suggested `within` over `over`.

Reply:
Personally I'd go with `window=10s`. Good thing we don't expose the option for window size yet, so we can save that discussion for another PR ;)